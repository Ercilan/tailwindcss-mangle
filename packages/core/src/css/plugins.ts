import type { ICssHandlerOptions } from '@/types'
import type { PluginCreator } from 'postcss'
import defu from 'defu'
import parser from 'postcss-selector-parser'

export type PostcssMangleTailwindcssPlugin = PluginCreator<ICssHandlerOptions>

const postcssPlugin = 'postcss-mangle-tailwindcss-plugin'

export function isVueScoped(s: parser.ClassName): boolean {
  if (s.parent) {
    const index = s.parent.nodes.indexOf(s)
    if (index > -1) {
      const nextNode = s.parent.nodes[index + 1]
      if (nextNode && nextNode.type === 'attribute' && nextNode.attribute.includes('data-v-')) {
        return true
      }
    }
  }
  return false
}

export const transformSelectorPostcssPlugin: PluginCreator<ICssHandlerOptions> = function (options) {
  const { ignoreVueScoped, ctx, id } = defu(options, {
    ignoreVueScoped: true,
  })

  const replaceMap = ctx.replaceMap

  return {
    postcssPlugin,
    Once(root) {
      root.walkRules((rule) => {
        parser((selectors) => {
          selectors.walkClasses((s) => {
            if (s.value && replaceMap && replaceMap.has(s.value)) {
              if (ignoreVueScoped && isVueScoped(s)) {
                return
              }
              const v = replaceMap.get(s.value)
              if (v) {
                if (ctx.isPreserveClass(s.value)) {
                  rule.cloneBefore()
                }
                // ctx.addToUsedBy(s.value, id)
                s.value = v
              }
            }
          })
        }).transformSync(rule, {
          lossless: false,
          updateSelector: true,
        })
      })
    },

  }
}
transformSelectorPostcssPlugin.postcss = true
