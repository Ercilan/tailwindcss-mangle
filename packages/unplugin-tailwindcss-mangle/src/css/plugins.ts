import { IClassGeneratorContextItem } from '@/types'
import type { PluginCreator } from 'postcss'
import parser from 'postcss-selector-parser'
export type PostcssMangleTailwindcssPlugin = PluginCreator<{
  newClassMap: Record<string, IClassGeneratorContextItem>
}>

const postcssPlugin = 'postcss-mangle-tailwindcss-plugin'

const postcssMangleTailwindcssPlugin: PostcssMangleTailwindcssPlugin = (options) => {
  // must set newClassMap options
  let newClassMap: Record<string, IClassGeneratorContextItem> = {}
  if (options) {
    if (options.newClassMap) {
      newClassMap = options.newClassMap
    }
  }

  return {
    postcssPlugin,
    Rule(rule, helper) {
      rule.selector = parser((selectors) => {
        selectors.walkClasses((s) => {
          if (s.value) {
            const hit = newClassMap[s.value]
            if (hit) {
              // vue scoped
              if (s.parent) {
                const idx = s.parent.nodes.indexOf(s)
                if (idx > -1) {
                  const nextNode = s.parent.nodes[idx + 1]
                  if (nextNode && nextNode.type === 'attribute' && nextNode.attribute.indexOf('data-v-') > -1) {
                    return
                  }
                }
              }
              s.value = hit.name
            }
          }
        })
      }).processSync(rule.selector)
    }
  }
}

postcssMangleTailwindcssPlugin.postcss = true

export { postcssMangleTailwindcssPlugin }
