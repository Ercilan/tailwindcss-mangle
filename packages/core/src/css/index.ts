import type { ICssHandlerOptions, IHandlerTransformResult } from '@/types'
import postcss from 'postcss'
import { transformSelectorPostcssPlugin } from './plugins'

export async function cssHandler(rawSource: string, options: ICssHandlerOptions): Promise<IHandlerTransformResult> {
  const acceptedPlugins = [transformSelectorPostcssPlugin(options)]
  const { id } = options
  try {
    const { css: code, map } = await postcss(acceptedPlugins).process(rawSource, {
      from: id,
      to: id,
    })
    return {
      code,
      // @ts-ignore
      map,
    }
  }
  catch (_error) {
    return {
      code: rawSource,
    }
  }
}
