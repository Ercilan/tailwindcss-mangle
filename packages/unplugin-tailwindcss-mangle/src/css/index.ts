import postcss from 'postcss'
import { postcssMangleTailwindcssPlugin } from './plugins'
import { ICssHandlerOptions } from '@/types'

export function cssHandler(rawSource: string, options: ICssHandlerOptions) {
  const acceptedPlugins = [postcssMangleTailwindcssPlugin(options)]
  return postcss(acceptedPlugins).process(rawSource).css
}
