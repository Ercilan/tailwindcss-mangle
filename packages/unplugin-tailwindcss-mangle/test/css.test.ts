import ClassGenerator from '../src/classGenerator'
import { cssHandler } from '../src/css/index'

describe('css', () => {
  it('vue scoped .gap-y-4[data-v-0f84999b]', () => {
    const classGenerator = new ClassGenerator()
    const runtimeSet = new Set<string>()
    runtimeSet.add('gap-y-4')
    classGenerator.generateClassName('gap-y-4')
    const testCase = `@media (min-width: 768px) {
      .gap-y-4[data-v-0f84999b] {
      }
    }`
    const css = cssHandler(testCase, {
      classGenerator,
      runtimeSet
    })
    expect(css).toMatchSnapshot()
  })
})
