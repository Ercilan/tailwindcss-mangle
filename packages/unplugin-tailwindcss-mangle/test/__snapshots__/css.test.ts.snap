// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`css vue scoped .gap-y-4[data-v-0f84999b] 1`] = `
"@media (min-width: 768px) {
      .tw-a[data-v-0f84999b] {
      }
    }"
`;
